File: 70.climbing-stairs.java
/*
 * @lc app=leetcode id=70 lang=java
 *
 * [70] Climbing Stairs
 */

// @lc code=start
class Solution {
    public int climbStairs(int n) {
        if (n <= 2) { // Trả về n nếu n nhỏ hơn hoặc bằng 2
            return n;
        }
        int[] dp = new int[n + 1]; // Tạo mảng dp để lưu số cách leo cầu thang
        dp[1] = 1; // Có 1 cách để leo 1 bậc
        dp[2] = 2; // Có 2 cách để leo 2 bậc
        for (int i = 3; i <= n; i++) {
            dp[i] = dp[i - 1] + dp[i - 2]; // Công thức quy hoạch động
        }
        return dp[n]; // Trả về số cách leo n bậc
    }
}
// @lc code=end

File: 70.climbing-stairs.py
#
# @lc app=leetcode id=70 lang=python3
#
# [70] Climbing Stairs
#

# @lc code=start
class Solution:
    def climbStairs(self, n: int) -> int:
        if n <= 2: # trường hợp cơ bản
            return n # trả về n
        first = 1 # số cách để lên bậc 1
        second = 2 # số cách để lên bậc 2
        for i in range(3, n + 1): # lặp từ bậc 3 đến bậc n
            third = first + second # số cách để lên bậc i là tổng số cách để lên bậc i-1 và i-2
            first = second # cập nhật số cách để lên bậc i-1
            second = third # cập nhật số cách để lên bậc i-2
        return second # trả về số cách để lên bậc n
# @lc code=end

